patchwork-fml/src/main/java/net/minecraftforge/forgespi/language/ModFileScanData.java

```diff
@@ -0,0 +1,239 @@
+			Gson gson = new Gson();
+			this.annotationStorage = gson.fromJson(
+					new InputStreamReader(outputStream), AnnotationStorage.class
+			);
+			annotationData = this.annotationStorage.entries.stream()
+					.map(ModFileScanData::getAnnotationData)
+					.collect(Collectors.toSet());
+		} catch (IOException e) {
+			LOGGER.error(String.format(
+					"Could not read annotations from %s %s (loaded from %s)",
+					modid, annotationJsonPath, modContainer.getRootPath()
+			));
+			e.printStackTrace();
+		}
+	}
+
+	public Set<AnnotationData> getAnnotations() {
+		if (!initialized) {
+			init();
+		}
+
+		return annotationData;
+	}
+
+	private static AnnotationData getAnnotationData(AnnotationStorage.Entry entry) {
+		Type annotationType = Type.getType(entry.annotationType);
+		Type targetInType = Type.getType("L" + entry.targetInClass + ";");
+		return new AnnotationData(
+				annotationType, entry.targetType, targetInType, entry.target
+		);
+	}
+
+	public static class AnnotationData {
+		private final Type annotationType;
+		private final ElementType targetType;
+		private final Type clazz;
+		private final String memberName;
+
+		//lazy evaluated
+		private Map<String, Object> annotationData;
+
+		public AnnotationData(
+				final Type annotationType, final ElementType targetType,
+				final Type clazz, final String memberName
+		) {
+			this.annotationType = annotationType;
+			this.targetType = targetType;
+			this.clazz = clazz;
+			this.memberName = memberName;
+		}
+
+		public Type getAnnotationType() {
+			return annotationType;
+		}
+
+		public ElementType getTargetType() {
+			return targetType;
+		}
+
+		public Type getClassType() {
+			return clazz;
+		}
+
+		public String getMemberName() {
+			return memberName;
+		}
+
+		public Map<String, Object> getAnnotationData() {
+			if (annotationData == null) {
+				initAnnotationData();
+			}
+
+			return annotationData;
+		}
+
+		private void initAnnotationData() {
+			annotationData = new HashMap<>();
+
+			try {
+				// TODO: This *may* load classes in the wrong order, but it shouldn't be an issue
+				Class<?> clazzObj = Class.forName(clazz.getClassName());
+				Class annotationType = Class.forName(this.annotationType.getClassName());
+				Annotation annotationObject = getAnnotationObject(clazzObj, annotationType);
+
+				if (annotationObject == null) {
+					LOGGER.error(String.format("Cannot fetch annotation object %s %s %s %s",
+							annotationType, targetType, clazz, memberName
+					));
+					return;
+				}
+
+				Method[] argMethods = annotationObject.getClass().getDeclaredMethods();
+
+				for (Method argMethod : argMethods) {
+					if (isArgumentMethod(argMethod)) {
+						annotationData.put(
+								argMethod.getName(),
+								argMethod.invoke(annotationObject)
+						);
+					}
+				}
+			} catch (Throwable e) {
+				LOGGER.catching(e);
+			}
+		}
+
+		private static boolean isArgumentMethod(Method method) {
```

Comment:
This name is confusing... what do you mean by an "argument method"?

Reply:
argument method is the annotation object's method to get the annotation argument